### PR TITLE
Fixing Flake with InitialUnavailable test by increasing failover timeout

### PR DESCRIPTION
**Pull request by donnadionne**
Currently, there are 2 backends that could will come up after the
initial unavailable.

Even tho the 2 backends have different prioirties, WaitForBackends will
simply send RPCs and then check the interested backend.  It is possible
for a RPC to get through to either backend in the process.

This is difficult to repro but I did manage to repro allowing 3 new backends
to increase the probability:
I0913 23:03:58.281209382      20 xds_end2end_test.cc:4503]   DONNAAA backend 0 has 0
I0913 23:03:58.283252494      20 xds_end2end_test.cc:4503]   DONNAAA backend 1 has 1
I0913 23:03:58.283321621      20 xds_end2end_test.cc:4503]   DONNAAA backend 2 has 0
I0913 23:03:58.283390738      20 xds_end2end_test.cc:4503]   DONNAAA backend 3 has 1

To make this test unflaky, let's just only bring up 1 backend after the
initial unavailable so that the behaviour will be deterministic.

@nicolasnoble

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/24147)

### CODE REVIEW

**Comment by markdroth** [comment]
> Even tho the 2 backends have different prioirties, WaitForBackends will simply send RPCs and then check the interested backend. It is possible for a RPC to get through to either backend in the process.

I don't believe that's correct.  The backends have different priorities, which means that it will try to connect to the higher priority one first.  It will only try the lower priority one if the higher priority one is unreachable, which should never be the case here.

Do you have evidence that the test is actually sending to backend 3 instead of backend 2?  I'd like to understand why that would be the case.

**Review by donnadionne — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/24147)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @markdroth)

___
*[test/cpp/end2end/xds_end2end_test.cc, line 4369 at r2](https://reviewable.io/reviews/grpc/grpc/24147#-MHEKfdy845l7my_CHrN:-MHEKfdy845l7my_CHrO:brvbvql) ([raw file](https://github.com/grpc/grpc/blob/3964d9a4f0f91cbb35f1f964868ad42ba12c3de9/test/cpp/end2end/xds_end2end_test.cc#L4369)):*
> ```cpp
>     BasicTest::SetUp();
>     ResetStub(500, "");
>   }
> ```

This is the only line of change I intend to submit.  The rest of the changes are just so that the problem is more easily reproducible but will not be included in the final submit.
___
*[test/cpp/end2end/xds_end2end_test.cc, line 4479 at r2](https://reviewable.io/reviews/grpc/grpc/24147#-MHEKr2I2jwtExmH3WNo:-MHEKr2I2jwtExmH3WNp:bi9dzmc) ([raw file](https://github.com/grpc/grpc/blob/3964d9a4f0f91cbb35f1f964868ad42ba12c3de9/test/cpp/end2end/xds_end2end_test.cc#L4479)):*
> ```cpp
>   SetNextResolutionForLbChannelAllBalancers();
>   AdsServiceImpl::EdsResourceArgs args({
>       {"locality0", GetBackendPorts(0, 1), kDefaultLocalityWeight, 0},
> ```

The test change below (which will not be submitted) makes it such that the second update will contain 3 backends with available priority instead of 2 backends.  

Because the 3 priorities differ, rpc should only always go to 1 backend consistently (the correct behaviour)

Having the 3 backends make the bug show up more easily: rpc did not always go to 1 backend only.

Once the fix (increasing failover timeout) is applied, rpc did always go to 1 backend (the correct behaviour)

So basically this change allowed me to repro the problem easily and test the fix easily; otherwise, I was not able to repro the problem with 1000-2000 runs.

I will ensure the test change is not submitted.  Just wanted to show you how I repro the problem

**Review by donnadionne — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/24147)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @markdroth)

___
*[test/cpp/end2end/xds_end2end_test.cc, line 4369 at r2](https://reviewable.io/reviews/grpc/grpc/24147#-MHEKfdy845l7my_CHrN:-MHEMV3VDFEE8Kv_Nmly:bfcfgti) ([raw file](https://github.com/grpc/grpc/blob/3964d9a4f0f91cbb35f1f964868ad42ba12c3de9/test/cpp/end2end/xds_end2end_test.cc#L4369)):*
<details><summary><i>Previously, donnadionne wrote…</i></summary><blockquote>

This is the only line of change I intend to submit.  The rest of the changes are just so that the problem is more easily reproducible but will not be included in the final submit.
</blockquote></details>

Once this fix is applied we no long got: 
I0914 19:15:38.606657832 40 priority.cc:658] [priority_lb 0x611000009b40] child child2 (0x613001702bc0): failover timer fired, reporting TRANSIENT_FAILURE

Instead we got:
I0915 00:58:23.603243565      34 priority.cc:622]            [priority_lb 0x6110000098c0] child child2 (0x613000012800): starting failover timer for 500 ms
...
I0915 00:58:23.628905183      40 priority.cc:638]            [priority_lb 0x6110000098c0] child child2 (0x613000012800): cancelling failover timer

File: test/cpp/end2end/xds_end2end_test.cc [hunk 1]
```diff
@@ -4498,9 +4496,11 @@ TEST_P(FailoverTest, UpdateInitialUnavailable) {
   do {
     CheckRpcSendFailure();
   } while (gpr_time_cmp(gpr_now(GPR_CLOCK_REALTIME), deadline) < 0);
-  WaitForBackend(2, false);
+  WaitForBackend(1, false);
   for (size_t i = 0; i < 4; ++i) {
-    if (i == 2) continue;
+    gpr_log(GPR_INFO, "backend %d request count %d", i,
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:4501]: This logging is not necessary.  If the request count is wrong, the expectation below will log the incorrect value.

**Review by markdroth — COMMENTED** [review]
I'm concerned that we needed to change the test in order to reproduce the problem.  The existing test should fail at least some percentage of the time, because otherwise it wouldn't show up as a flake.  Why can't we reproduce that?

If we can't reproduce the flake with the existing test, then how do we know that this will actually fix the problem?  It could actually turn out to be something completely different causing the problem.

Now, that having been said, I do think it's very likely that this will actually fix the problem, so I'm not going to block this if we can't reproduce it with the existing test.  But I'd feel a lot better about it if we could reliably reproduce it.

**Review by donnadionne — COMMENTED** [review]
i understand; this problem is difficult to repro, for example, this month of sept it occurred 5 times in all our runs and combinations.  i think we should put it in and monitor and expect the problem to not occur again?

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/24147#-:-MHHmgj4-XeQotPAyTNZ:bimtw8g)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @markdroth)

___
*[test/cpp/end2end/xds_end2end_test.cc, line 4501 at r2](https://reviewable.io/reviews/grpc/grpc/24147#-MHEKceK6f-lyPDTkmGW-r2-4501:-MHHn2jfCkOrvyyQULMP:bqeaelb) ([raw file](https://github.com/grpc/grpc/blob/3964d9a4f0f91cbb35f1f964868ad42ba12c3de9/test/cpp/end2end/xds_end2end_test.cc#L4501)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

This logging is not necessary.  If the request count is wrong, the expectation below will log the incorrect value.
</blockquote></details>

will be removed.

did have a question, the expectation will log the incorrect value, but it does not log the value of "i", the backend number which can be helpful.  not sure if we can do that.

**Review by donnadionne — COMMENTED** [review]
updated, sqaushed to just 1 fine fix.

> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/24147#-:-MHHsqJ8AXATAn6fWFjN:b-w6uiq7)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @markdroth)

**Review by markdroth — APPROVED** [review]
Okay.  It's disappointing that we can't be more sure, but I do think it's likely that this will fix it.

Thanks!

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/24147#-:-MHIE0wX0G2v0gnNz8Ym:b-r60cjn)* status: all files reviewed, 2 unresolved discussions (waiting on @donnadionne)

___
*[test/cpp/end2end/xds_end2end_test.cc, line 4501 at r2](https://reviewable.io/reviews/grpc/grpc/24147#-MHEKceK6f-lyPDTkmGW-r2-4501:-MHIDRom4a39vGPvuIQJ:b-xjwhoo) ([raw file](https://github.com/grpc/grpc/blob/3964d9a4f0f91cbb35f1f964868ad42ba12c3de9/test/cpp/end2end/xds_end2end_test.cc#L4501)):*
<details><summary><i>Previously, donnadionne wrote…</i></summary><blockquote>

will be removed.

did have a question, the expectation will log the incorrect value, but it does not log the value of "i", the backend number which can be helpful.  not sure if we can do that.
</blockquote></details>

You can say `EXPECT_EQ(...) << i;`, and it will log the value of `i` when the expectation fails.

That having been said, in this particular case, it would probably be a lot simpler to just get rid of the loop and directly check each value:

```
EXPECT_EQ(0U, backends_[0]->backend_service()->request_count());
EXPECT_EQ(0U, backends_[1]->backend_service()->request_count());
EXPECT_EQ(0U, backends_[3]->backend_service()->request_count());
```
